src/torchcodec/samplers/_implem.py

```diff
@@ -0,0 +1,62 @@
+from typing import List
+
+import torch
+
+from torchcodec.decoders import (  # TODO: move FrameBatch to torchcodec.FrameBatch?
+    FrameBatch,
+    SimpleVideoDecoder,
+)
+
+
+def clips_at_random_indices(
+    decoder: SimpleVideoDecoder,
+    *,
+    num_clips: int = 1,
+    num_frames_per_clip: int = 1,
+    num_indices_between_frames: int = 1,
+) -> List[FrameBatch]:
+    if num_clips <= 0:
+        raise ValueError(f"num_clips ({num_clips}) must be strictly positive")
+    if num_frames_per_clip <= 0:
+        raise ValueError(
+            f"num_frames_per_clip ({num_frames_per_clip}) must be strictly positive"
+        )
+    if num_indices_between_frames <= 0:
+        raise ValueError(
+            f"num_indices_between_frames ({num_indices_between_frames}) must be strictly positive"
+        )
+
+    # Determine the span of a clip, i.e. the number of frames (or indices)
+    # between the first and last frame in the clip, both included. This isn't
+    # the same as the number of frames in a clip!
+    # Example: f means a frame in the clip, x means a frame excluded from the clip
+    # num_frames_per_clip = 4
+    # num_indices_between_frames = 1, clip = ffff      , span = 4
+    # num_indices_between_frames = 2, clip = fxfxfxf   , span = 7
+    # num_indices_between_frames = 3, clip = fxxfxxfxxf, span = 10
+    clip_span = num_indices_between_frames * (num_frames_per_clip - 1) + 1
+
+    # TODO: We should probably not error.
+    if clip_span > len(decoder):
+        raise ValueError(
+            f"Clip span ({clip_span}) is larger than the number of frames ({len(decoder)})"
+        )
+
+    last_clip_start_index = len(decoder) - clip_span
+    clip_start_indices = torch.randint(
+        low=0, high=last_clip_start_index + 1, size=(num_clips,)
+    )
+
+    # TODO: This is inefficient as we are potentially seeking backwards.
+    # We should sort by clip start before querying, and re-shuffle.
+    # Note: we may still have to seek backwards if we have overlapping clips.
```

Comment:
@ahmadsharif1 @scotts 

This is the simplest implementation that works, but of course it's not efficient. I'm seeing 2 potential ways of improving irt:

1. What's described in the comment above i.e. sort the clip start, call `get_frames_at`, and re-shuffle the clips by clip start. This makes sure clips are decoded in order, but we may still have to seek backwards if a clip ends *after* the next one starts. I think this can be implemented fairly easily with the current APIs that we have.

2. flatten all frame indices across all clips, sort them, decode all the frames in sorted order, re-order the frames to their original shuffled order, and then re-assemble the frames into clips (this is easy since we know all clips are of the same size). I suspect this is the fastest and most efficient. I think this can be done with our current APIs, but it may benefit from a new `decode_frames_at_those_indices(indices)` API.

LMK you thoughts on the proposed strategies, and if you have any alternative suggestion.

Reply:
I think Option 1 is a reasonable optimization that is quick and intuitive enough to do for a first implementation.

Option 2 is something we can do if we discover Option 1 is too slow. The C++ API will probably be generally useful outside of our sampler - that is, for users who have their own bespoke sampler. (I have seen one instance of this in the wild.) I believe we would require the decoder to be scanned for this behavior, of course deferring to @ahmadsharif1's understanding.

Note that `SimpleVideoDecoder` always scans the file - although I think our future decoders may not always scan the file, so we should still consider this case. (Getting a proposal up for `SingleStreamVideoDecoder` is one of my TODOs for this week.)

Reply:
For the fastest implementation, you would need to flatten, sort and dedupe frames, then ask the decoder to decode in that order, and then fill in the FrameBatch.

I think torch.rand can return duplicate values and even without duplicate frame starts, based on dilation you can get duplicate frames. You want to decode them just once, for the fastest performance (since we don't cache them in the C++ code).

Seeking backwards is slow.

I think even getting frame count reliably is hard without scanning the file. At least I don't know of a reliable way that works 100% of the time without scanning.

However if the frame count is in the header and you want to trust it for a fast sampler you could do that from the core API.

Reply:
We decided to go with option 1 for now as it's simpler to implement, and immediately follow-up with option 2.